math/binary/sqrt.go

```diff
@@ -0,0 +1,26 @@
+// sqrt.go
+// description: Square root calculation
+// details:
+// Calculating the square root using binary operations and a magic number 0x5f3759df [See more](https://en.wikipedia.org/wiki/Fast_inverse_square_root)
+// author(s) [red_byte](https://github.com/i-redbyte)
+// see sqrt_test.go
+
+package binary
+
+import (
+	"math"
+)
+
+const threeHalves = 1.5
+
+func Sqrt(number float32) float64 {
+	var halfHumber, y float32
+	halfHumber = number * 0.5
+	z := math.Float32bits(number) // floating point bit level hacking
+	z = 0x5f3759df - (z >> 1)     // Newton's approximation
+	y = math.Float32frombits(z)
+	for i := 0; i < 3; i++ {
+		y = y * (threeHalves - (halfHumber * y * y))
+	}
+	return math.RoundToEven(float64(1 / y))
```

Comment:
You are converting a 32 bit to a 64 bit? Shouldn't it adhere to what the input numbers type?

Reply:
If you're using the float64 then compare it with the `math` package's sqrt as well.

Reply:
accepts only float64 without rounding problems begin.

Reply:
math.RoundToEven  accepts only float64 and without rounding, problems begin. I don't know how to solve this problem nicely

tests have BenchmarkMathSqrt measurement

Reply:
and if you use the float64 input parameter, then the cast
z := math.Float64bits(number)
z = 0x5f3759df - (z >> 1)

it works incorrectly

Reply:
> math.RoundToEven accepts only float64 and without rounding, problems begin. I don't know how to solve this problem nicely

I think people looking at this algorithm would know the peculiarity of `float`s, so rounding is not necessary I think. The equality of floats is almost always done like so `Abs(actualValue - calculatedValue) < epsilon` for some small enough epsilon (usually depends on the threshold of float size (32 bit or 64 bit) so it makes sense to not have to round the number.

Reply:
Also, regarding the benchmark, can you make sure that your architecture is one of these mentioned [here](https://cs.opensource.google/go/go/+/refs/tags/go1.17.8:src/math/sqrt_asm.go;drc=refs%2Ftags%2Fgo1.17.8;l=5)?

Reply:
I tested on two devices(arm64), this architecture is on the list
1)
goos: darwin
goarch: amd64
cpu: Intel(R) Core(TM) i7-4578U CPU @ 3.00GHz
BenchmarkSquareRootCalculation
BenchmarkSquareRootCalculation-4   	1000000000	         0.3576 ns/op
PASS

2)

goos: darwin
goarch: amd64
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkSquareRootCalculation
BenchmarkSquareRootCalculation-16      1000000000           0.2569 ns/op
PASS